File: gtci/05-cyclic-sort/cpp/1-find-the-missing-number.cpp
#include <cassert>
#include <iostream>
#include <utility>
#include <vector>

using namespace std;

class Solution {
   public:
    static int findMissingNumber(vector<int> &&nums) {
        size_t i = 0;
        while (i < nums.size()) {
            int num = nums[i];

            if (num < nums.size() && i != num) {
                std::swap(nums[i], nums[num]);
            } else {
                i++;
            }
        }

        for (i = 0; i < nums.size(); i++) {
            if (i != nums[i]) return i;
        }

        return nums.size();
    }
};

int main() {
    Solution sol;

    assert(sol.findMissingNumber(vector<int>{3, 0, 1}) == 2);
    assert(sol.findMissingNumber(vector<int>{0, 1}) == 2);
    assert(sol.findMissingNumber(vector<int>{9, 6, 4, 2, 3, 5, 7, 0, 1}) == 8);

    cout << "All test cases passed." << endl;

    return 0;
}

File: gtci/05-cyclic-sort/python3/1-find-the-missing-number.py
class Solution:
    def findMissingNumber(self, nums):
        i = 0
        while i < len(nums):
            num = nums[i]

            if num < len(nums) and num != i:
                nums[i], nums[num] = nums[num], nums[i]
            else:
                i += 1

        for i in range(len(nums)):
            if i != nums[i]:
                return i

        return len(nums)


if __name__ == "__main__":
    sol = Solution()

    assert sol.findMissingNumber([3, 0, 1]) == 2
    assert sol.findMissingNumber([0, 1]) == 2
    assert sol.findMissingNumber([9, 6, 4, 2, 3, 5, 7, 0, 1]) == 8

    print("All test cases passed.")
